### ISSUE TITLE
Prepare and release 1.6.0

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctorj#702]
The time has come to make 1.6.0 an official release! :tada: This release will be based on Asciidoctor 1.5.8.

As a first step, the code in master will be branched as asciidoctorj-1.5.x and the code in the asciidoctorj-1.6.0 branch will be folded into master.

The reason we're planning to release 1.6.0 and not jump directly to 2.0.0 is to find out whether any adjustments need to be made to the API breakages before we transition to semantic versioning, which will start with 2.0.0.

Once this release is out, all new users should start with 1.6.0. Veteran users should begin testing this new version to help us sort out kinks and get ready for the transition to 2.0.0. The AsciidoctorJ 2.0.0 release will then follow and be based on Asciidoctor 2.0.0.

1.6.0 is considered a major release (despite the version number). It's substantially different from the 1.5.x releases and thus not binary compatible. This was necessary to resolve long-standing issues in the 1.5.x series. Specifically, this release introduces a new way to bridge Java and Ruby objects which is much more stable and reliable. This change will be most noticeable to extension writers and to those building documents programmatically. It also introduces an AST (document model) that's more well thought out and semantic. The core jar has be partitioned into an API and implementation to further bolster these AST classes and make it easier to see where the API ends and the implementation begins. More details about these changes will be provided in a blog post that introduces this new version.

We'll likely publish a release candidate before the 1.6.0 to ensure the release process is still stable as the build has undergone some changes recently. We're also waiting on the Asciidoctor 1.5.8 release.

* [x] Create the asciidoctorj-1.5.x branch from master
* [x] Merge asciidoctorj-1.6.0 into master
* [x] Wait for Asciidoctor 1.5.8 release
* [x] Write blog post introducing the release (state that this is a major version, emphasize that it's not binary compatible w/ 1.5.x, identify which APIs have changed, provide advice for migrating)
* [x] Publish a release candidate

**Comment by abelsromero** [comment]
Just reporting, there's an issue with latest JRuby versions (`9.1.17.0` & `9.2.0.0`) that prevents some features like remote include in Java11. New version to be release soon fixes it (https://github.com/jruby/jruby/issues/5374#issuecomment-431604262).

**Comment by robertpanzer** [comment]
I created a branch `asciidoctorj-1.5.x` that now points to the head of the 1.5.x releases.
Note that there was such a branch in the past which wasn't of much use, so if you can't fetch it because you already have a conflicting branch, please delete that.

Then I merged asciidoctorj-1.6.0 into master effectively taking asciidoctorj-1.6.0 only.
You should be able to fast-forward from master without interruptions.
Few things that got lost because they were only available in master yet:
- publish to sdkman
- publish to chocolatey
- .gitattributes fixing line endings for asciidoc test files

I'll redo this over the next days.

So if you want to create PRs for the current 1.5 line please create PRs to `asciidoctorj-1.5.x`. 
For 1.6.0 create PRs for `master`.
The asciidoctorj-1.6.0 branch should be dead now.

**Comment by mojavelinux** [comment]
I'm updating the description with a checklist so we can track where we are.

**Comment by mojavelinux** [comment]
@robertpanzer I think your list is done now, right?

**Comment by robertpanzer** [comment]
Yes. 

The only thing right now is the concurrent ruby gem that I have to fix on master as well. 
Also I am still wondering if I could continue the split into api and impl so that we can make the api available to AsciidoctorG, but maybe that’s sth to do later. 
(I am really keen though to get my fingers on this as well 😃 it’s just a matter of time...)

**Comment by mojavelinux** [comment]
Here's a comment a wrote in another thread explaining the 1.6.0 and 2.0.0 releases. This is the starting point for the mailinglist or blog post we'll need to make:

---

AsciidoctorJ 1.6.0 is a major release for AsciidoctorJ to help the transition to the subsequent AsciidoctorJ 2.0.0 major release. You can think of it as early access to AsciidoctorJ 2.0.0. The reason we're making it is to get feedback on the major API changes and so we can make adjustments to get AsciidoctorJ 2.0.0 right. Trying to do this all in 2.0.0 would just be too much.

I expect the bulk of users will transition directly from AsciidoctorJ 1.5.8 to 2.0.0. But 1.6.0 is there to help them get ready.

**Comment by ysb33r** [comment]
I am a bit confused: asciidoctorj 1.5.8 becomes 1.6.0 and asciidoctorj 1.6.0-alpha.X becomes 2.0.0.-alpha.X ?

**Comment by mojavelinux** [comment]
AsciidoctorJ 1.6.0-alpha.X becomes 1.6.0, a stable major release. But it will use Asciidoctor 1.5.8 under the covers. The focus is on solidifying the existing API changes in AsciidoctorJ (not changes in Asciidoctor itself).

AsciidoctorJ 2.0.0 will also be a major release. There are two key differences from 1.6.0. First, it marks a transition to semantic versioning. Second, it upgrades to Asciidoctor 2.0.0. However, we'll likely stop linking the version of AsciidoctorJ to Aciidoctor after that release. I'll provide information about that in a follow-up comment.

**Comment by ysb33r** [comment]
Got it! There is no change for the development of the Gradle plugin in this case. The 1.5.x series will keep tracking AsciidoctorJ 1.5 releases and will probably just work with AsciidoctorJ 1.6.0. If the latter is the case we'll just default to 1.6.

The Gradle plugin 2.x.alpha.X series already uses the AsciidoctorJ 1.6.0-alpha.X releases. When AsciidoctorJ 1.6.0 is fully released, we'll be close to releasing 2.0 of the Gradle plugin.

@abelsromero will have to say what his strategy will be for the Maven plugin.

**Comment by mojavelinux** [comment]
Great! Exactly the feedback we're looking for!

**Comment by abelsromero** [comment]
Current 1.5.x is compatible with Asciidoctorj 1.6.x alphas (need to test RC) and if some issue appears my idea is to fix it as a minor version, but the default version will still be 1.5.7.1.

About 2.0.0 I have to say I am not so keen of the idea of diverging the versions from the gem. 
For me is very convenient and simplifies explaining the Asciidoctor stack on Java and issue diagnosis. For example, it happens sometimes that one as to reproduce things directly with the ruby gem and having to guess the right version adds another complexity layer for those not familiar.
Also, what will happen when the gem gets a an eventual 2.0.x release?

**Comment by mojavelinux** [comment]
I'll let @Mogztter chime in on the first part since he provided the key defense for not locking them together.

I'll answer this question:

> what will happen when the gem gets a an eventual 2.0.x release?

Asciidoctor, AsciidoctorJ, and Asciidoctor.js are all going to move to 2.0.0 at the same time so we can start with a clean slate. Only after that will they diverge.

In a few days, I'll post these plans to the mailinglist, where we'll be able to discuss at greater length. My goal is to have the 2.0.0 release out before the end of the year.

**Comment by abelsromero** [comment]
Now it's clear, I got the Asciidoctor gem versioning wrong, mistake totally on my side.

> In a few days, I'll post these plans to the mailinglist, where we'll be able to discuss at greater length. My goal is to have the 2.0.0 release out before the end of the year.

When that's ready I'll create an issue to track all changes for 2.0.x, no rush.
So far I am still tagging things in the maven plugin as v 1.6.0.

**Comment by Mogztter** [comment]
> About 2.0.0 I have to say I am not so keen of the idea of diverging the versions from the gem.

I think that AsciidoctorJ has reached the point where the project is so much more than just the Asciidoctor gem available on the JVM.
The project has a complete API, extensions and also custom features to provide better integration with the Java ecosystem.

Semantic versioning is all about shared expectations.
Let's say we want to break an API in AsciidoctorJ for instance to ease the integration in Maven 5.
I (as a user of AsciidoctorJ) expect this change to happen in major version of AsciidoctorJ.
But at the same time Asciidoctor.js and/or Asciidoctor Ruby don't have plan to make a breaking change. So what now ?

Should we release a major version of Asciidoctor.js and Asciidoctor Ruby even if the major version does not contain any breaking change ? If we do that we broke our promise. Users will carefully upgrade to the major version, take time to read the changelog to eventually find out that well... in this version we didn't break anything... A user could then ask us "why are you releasing a major version then ? You are not using semantic versioning!" (and he will be right).

The other way around, is to wait for Asciidoctor.js and Asciidoctor Ruby to break something but now everyone is tied and every project has to wait two others projects.
I would argue that the pace at which each language/ecosystem/project evolves is not the same. And that's a good thing, we should just embrace that fact.

> For me is very convenient and simplifies explaining the Asciidoctor stack on Java and issue diagnosis. For example, it happens sometimes that one as to reproduce things directly with the ruby gem and having to guess the right version adds another complexity layer for those not familiar.

@abelsromero To work around this specific issue you can ask the user to provide the version using the `Asciidoctor.asciidoctorVersion()` API. You can also state in each release note of AsciidoctorJ which version of Asciidoctor core is used (I'm already doing it in Asciidoctor.js https://github.com/asciidoctor/asciidoctor.js/releases/tag/v1.5.7) 

https://github.com/asciidoctor/asciidoctorj/blob/e9b81c6a16d2c176377d84b57c5809243739f995/asciidoctorj-core/src/main/java/org/asciidoctor/internal/JRubyAsciidoctor.java#L369-L371

Even if the user doesn't provide the Asciidoctor core version, it will be straightforward to know that version X of AsciidoctorJ is using version Y of Asciidoctor Ruby.

Also I think that users don't really care about aligned version. They care about features and bug fixes. If I find a bug in AsciidoctorJ, I (as a user) don't really care if the bug is in Asciidoctor core or in AsciidoctorJ itself. I just want the bug to be fixed so I can upgrade to the next patch version. Maybe it will involve a bug fix in Asciidoctor core, maybe not.
It's true that they also care about the AsciiDoc syntax they can use and that why it's important to explicitly state which features is available.

**Comment by robertpanzer** [comment]
Do we have any open points now besides writing the blog post?
I tested 1.6.0-RC.2 with asciidoctor-maven-plugin 1.5.6 and that seems to work.
I'll also test if extensions work, but I don't see any reasons why it should not work.

**Comment by ysb33r** [comment]
I know extensions work with 2.0-alpha.4 of the Gradle plugin

**Comment by mojavelinux** [comment]
The release post is now ready. See https://github.com/asciidoctor/asciidoctor.org/pull/828.

**Comment by ahus1** [comment]
@robertpanzer - I suppose this ticket can be closed as 1.6 has been announced some time ago.

**Comment by robertpanzer** [comment]
Thanks for the reminder! We’re definitely past that
